File: packages/shared-contracts/src/draw-config.ts
import { z } from "zod";
import { DrawConfigIdSchema, EventIdSchema } from "./ids.js";

/** Textos voltados ao público / telas externas (opcional; pt-BR típico no conteúdo). */
export const DrawPublicCopyMvpSchema = z.object({
  headline: z.string().max(200).optional(),
  audience_instructions: z.string().max(500).optional(),
  result_label: z.string().max(120).optional(),
});

export type DrawPublicCopyMvp = z.infer<typeof DrawPublicCopyMvpSchema>;

/** Intervalo inclusivo para sorteio numérico (`number_range` e modo `subset`). */
export const DrawNumberRangeSchema = z.object({
  min: z.number().int(),
  max: z.number().int(),
});

export type DrawNumberRange = z.infer<typeof DrawNumberRangeSchema>;

export const DrawTypeSchema = z.enum(["number_range", "attendee_pool"]);

export type DrawType = z.infer<typeof DrawTypeSchema>;

export const DrawPoolModeSchema = z.enum(["full_range", "subset"]);

export type DrawPoolMode = z.infer<typeof DrawPoolModeSchema>;

export const DrawVisualProfileSchema = z.enum(["premium", "classic", "pulsing"]);

export type DrawVisualProfile = z.infer<typeof DrawVisualProfileSchema>;

/** Som, visual e overrides da animação (Player `drawEngine` / CSS). Campos opcionais no pack. */
export const DrawPresentationPartialSchema = z
  .object({
    sound_enabled: z.boolean().optional(),
    visual_profile: DrawVisualProfileSchema.optional(),
    tick_count: z.number().int().min(10).max(120).optional(),
    total_duration_ms: z.number().int().min(500).max(60000).optional(),
    freeze_before_final_ms: z.number().int().min(0).max(5000).optional(),
  })
  .optional();

export type DrawPresentationPartial = z.infer<typeof DrawPresentationPartialSchema>;

/** Referência para QR / inscrições (congelado no pack). */
export const DrawRegistrationRefSchema = z.object({
  join_url_template: z.string().max(1024).optional(),
  public_token: z.string().max(128).optional(),
});

export type DrawRegistrationRef = z.infer<typeof DrawRegistrationRefSchema>;

/**
 * DrawConfig owned by the event; scenes reference optionally (EVENT_EDITOR_FEATURE_SPEC).
 * `number_range` opcional: se ausente, o Player MVP pode usar intervalo padrão.
 */
export const DrawConfigContractSchema = z
  .object({
    draw_config_id: DrawConfigIdSchema,
    event_id: EventIdSchema,
    name: z.string().min(1).max(256),
    max_winners: z.number().int().min(1).max(999).default(1),
    notes: z.string().max(2000).nullish(),
    enabled: z.boolean().default(true),
    draw_type: DrawTypeSchema.default("number_range"),
    number_range: DrawNumberRangeSchema.optional(),
    pool_mode: DrawPoolModeSchema.default("full_range"),
    eligible_numbers: z.array(z.number().int()).max(50000).optional(),
    remove_winner_from_pool: z.boolean().default(true),
    prizes: z.array(z.string().min(1).max(256)).max(99).optional(),
    public_copy: DrawPublicCopyMvpSchema.nullish(),
    draw_presentation: DrawPresentationPartialSchema,
    registration: DrawRegistrationRefSchema.optional(),
  })
  .superRefine((cfg, ctx) => {
    if (cfg.number_range && cfg.number_range.max < cfg.number_range.min) {
      ctx.addIssue({
        code: "custom",
        path: ["number_range", "max"],
        message: "number_range.max must be >= number_range.min",
      });
    }
    if (cfg.pool_mode === "subset" && (!cfg.eligible_numbers || cfg.eligible_numbers.length === 0)) {
      ctx.addIssue({
        code: "custom",
        path: ["eligible_numbers"],
        message: "pool_mode subset requires eligible_numbers (non-empty).",
      });
    }
  });

export type DrawConfigContract = z.infer<typeof DrawConfigContractSchema>;

File: apps/cloud-api/src/telaflow_cloud_api/domain/draw_config.py
"""Sorteio no âmbito do evento — alinhado a shared-contracts (draw-config)."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, model_validator

from telaflow_cloud_api.domain.draw_public_copy import DrawPublicCopy

_ID_PATTERN = r"^[a-zA-Z0-9_-]+$"


class NumberRange(BaseModel):
    """Intervalo inclusivo para `draw_type=number_range` (e modo `subset`)."""

    model_config = ConfigDict(extra="forbid")

    min: int
    max: int


class DrawPresentationPartial(BaseModel):
    model_config = ConfigDict(extra="forbid")

    sound_enabled: bool | None = None
    visual_profile: Literal["premium", "classic", "pulsing"] | None = None
    tick_count: int | None = Field(default=None, ge=10, le=120)
    total_duration_ms: int | None = Field(default=None, ge=500, le=60000)
    freeze_before_final_ms: int | None = Field(default=None, ge=0, le=5000)


class DrawRegistrationRef(BaseModel):
    model_config = ConfigDict(extra="forbid")

    join_url_template: str | None = Field(default=None, max_length=1024)
    public_token: str | None = Field(default=None, max_length=128)


class DrawConfig(BaseModel):
    """DrawConfig pertence ao evento; a scene só referencia opcionalmente."""

    draw_config_id: str = Field(..., min_length=12, max_length=64, pattern=_ID_PATTERN)
    event_id: str = Field(..., min_length=12, max_length=64, pattern=_ID_PATTERN)
    name: str = Field(..., min_length=1, max_length=256)
    max_winners: int = Field(default=1, ge=1, le=999)
    notes: str | None = Field(default=None, max_length=2000)
    enabled: bool = True
    draw_type: Literal["number_range", "attendee_pool"] = "number_range"
    number_range: NumberRange | None = None
    pool_mode: Literal["full_range", "subset"] = "full_range"
    eligible_numbers: list[int] | None = Field(default=None, max_length=50000)
    remove_winner_from_pool: bool = True
    prizes: list[str] | None = Field(default=None, max_length=99)
    public_copy: DrawPublicCopy | None = None
    draw_presentation: DrawPresentationPartial | None = None
    registration: DrawRegistrationRef | None = None

    @model_validator(mode="after")
    def _consistency(self) -> DrawConfig:
        if self.number_range is not None and self.number_range.max < self.number_range.min:
            raise ValueError("number_range.max must be >= number_range.min")
        if self.pool_mode == "subset" and (not self.eligible_numbers or len(self.eligible_numbers) == 0):
            raise ValueError("pool_mode subset requires eligible_numbers (non-empty)")
        return self


class DrawConfigCreate(BaseModel):
    """Criação — `draw_config_id` e `event_id` definidos pela rota."""

    name: str = Field(..., min_length=1, max_length=256)
    max_winners: int = Field(default=1, ge=1, le=999)
    notes: str | None = Field(default=None, max_length=2000)
    enabled: bool = True
    draw_type: Literal["number_range", "attendee_pool"] = "number_range"
    number_range: NumberRange | None = None
    pool_mode: Literal["full_range", "subset"] = "full_range"
    eligible_numbers: list[int] | None = Field(default=None, max_length=50000)
    remove_winner_from_pool: bool = True
    prizes: list[str] | None = Field(default=None, max_length=99)
    public_copy: DrawPublicCopy | None = None
    draw_presentation: DrawPresentationPartial | None = None
    registration: DrawRegistrationRef | None = None

    @model_validator(mode="after")
    def _create_consistency(self) -> DrawConfigCreate:
        if self.number_range is not None and self.number_range.max < self.number_range.min:
            raise ValueError("number_range.max must be >= number_range.min")
        if self.pool_mode == "subset" and (not self.eligible_numbers or len(self.eligible_numbers) == 0):
            raise ValueError("pool_mode subset requires eligible_numbers (non-empty)")
        return self


class DrawConfigUpdate(BaseModel):
    """Atualização parcial (PATCH)."""

    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=256)
    max_winners: int | None = Field(default=None, ge=1, le=999)
    notes: str | None = None
    enabled: bool | None = None
    draw_type: Literal["number_range", "attendee_pool"] | None = None
    number_range: NumberRange | None = None
    pool_mode: Literal["full_range", "subset"] | None = None
    eligible_numbers: list[int] | None = None
    remove_winner_from_pool: bool | None = None
    prizes: list[str] | None = None
    public_copy: DrawPublicCopy | None = None
    draw_presentation: DrawPresentationPartial | None = None
    registration: DrawRegistrationRef | None = None
